File: algorithms/2_medium/T/01415_the_k_th_lexicographical_string_of_all_happy_strings_of_length_n/dfs.c
#define VALID_CHARS_LEN 3

const static int valid_chars[VALID_CHARS_LEN] = {'a', 'b', 'c'};

static void dfs(char* temp, const int i, const int n, int* cnt, const int target)
{
    if (*cnt == target) {
        return;
    }

    if (i == n) {
        *cnt += 1;
        return;
    }

    for (int idx = 0; idx < VALID_CHARS_LEN; idx += 1) {
        if (i == 0 || temp[i - 1] != valid_chars[idx]) {
            temp[i] = valid_chars[idx];
            dfs(temp, i + 1, n, cnt, target);
        }

        if (*cnt == target) {
            return;
        }
    }

}

char* getHappyString(int n, int k)
{
    assert(n >= 1 && k >= 1 && n <= 10 && k <= 100);
    int cnt = 0;
    char* temp = malloc((size_t)(n + 1) * sizeof(*temp));
    assert(temp != NULL);
    dfs(temp, 0, n, &cnt, k);
    if (cnt == k) {
        temp[n] = '\0';
        return temp;
    }
    temp[0] = '\0';
    return temp;
}

File: algorithms/2_medium/T/01415_the_k_th_lexicographical_string_of_all_happy_strings_of_length_n/dfs.rs
impl Solution {
    const VALID_BYTES: [u8; 3] = [b'a', b'b', b'c'];

    fn dfs(temp: &mut [u8], i: usize, cnt: &mut i32, target: i32) {
        if *cnt == target {
            return;
        }

        if i == temp.len() {
            *cnt += 1;
            return;
        }

        for byte in Self::VALID_BYTES {
            if i == 0 || temp[i - 1] != byte {
                temp[i] = byte;
                Self::dfs(temp, i + 1, cnt, target);
            }

            if *cnt == target {
                return;
            }
        }

    }

    pub fn get_happy_string(n: i32, k: i32) -> String {
        let mut cnt = 0_i32;
        let mut temp = vec![b' '; n as usize];
        Self::dfs(&mut temp, 0, &mut cnt, k);
        if cnt == k {
            return String::from_utf8(temp).unwrap();
        }
        String::default()
    }
}
